python/vyos/config_mgmt.py

```diff
@@ -0,0 +1,664 @@
+            ct1 = self._get_config_tree_revision(rev2)
+            msg = f'No changes between revisions {rev2} and {rev1} configurations.\n'
+
+        if commands:
+            lines1 = ct1.to_commands().splitlines(keepends=True)
+            lines2 = ct2.to_commands().splitlines(keepends=True)
+        else:
+            lines1 = ct1.to_string().splitlines(keepends=True)
+            lines2 = ct2.to_string().splitlines(keepends=True)
+
+        out = ''
+        comp = unified_diff(lines1, lines2)
+        for line in comp:
+            if re.match(r'(\-\-)|(\+\+)|(@@)', line):
+                continue
+            out += line
+        if out:
+            msg = out
+
+        return msg, 0
+
+    def wrap_compare(self, options) -> Tuple[str,int]:
+        """Interface to vyatta-cfg-run: args collected as 'options' to parse
+        for compare.
+        """
+        cmnds = False
+        r1 = None
+        r2 = None
+        if 'commands' in options:
+            cmnds=True
+            options.remove('commands')
+        for i in options:
+            if not i.isnumeric():
+                options.remove(i)
+        if len(options) > 0:
+            r1 = int(options[0])
+        if len(options) > 1:
+            r2 = int(options[1])
+
+        return self.compare(commands=cmnds, rev1=r1, rev2=r2)
+
+    # Initialization and post-commit hooks for conf-mode
+    #
+    def initialize_revision(self):
+        """Initialize config archive, logrotate conf, and commit log.
+        """
+        mask = os.umask(0o002)
+        os.makedirs(archive_dir, exist_ok=True)
+
+        self._add_logrotate_conf()
+
+        if (not os.path.exists(commit_log_file) or
+            self._get_number_of_revisions() == 0):
+            user = self._get_user()
+            via = 'init'
+            comment = ''
+            self._add_log_entry(user, via, comment)
+            # add empty init config before boot-config load for revision
+            # and diff consistency
+            if self._archive_active_config():
+                self._update_archive()
+
+        os.umask(mask)
+
+    def commit_revision(self):
+        """Update commit log and rotate archived config.boot.
+
+        commit_revision is called in post-commit-hooks, if
+        ['commit-archive', 'commit-revisions'] is configured.
+        """
+        if os.getenv('IN_COMMIT_CONFIRM', ''):
+            self._new_log_entry(tmp_file=tmp_log_entry)
+            return
+
+        self._add_log_entry()
+
+        if self._archive_active_config():
+            self._update_archive()
+
+    def commit_archive(self):
+        """Upload config to remote archive.
+        """
+        hostname = self.hostname
+        t = datetime.now()
+        timestamp = t.strftime('%Y%m%d_%H%M%S')
+        remote_file = f'config.boot-{hostname}.{timestamp}'
+        source_address = self.source_address
+
+        for location in self.locations:
+            upload(archive_config_file, f'{location}/{remote_file}',
```

Comment:
One occasionally-requested feature is saving config revisions in a remote git repository. I think we may want to add an abstract "archive storage" interface right away and keep it to a single "URL upload" implementation for now, so that later these functions will be easy to adjust to instantiating a storage object:

Pseudocode:
```
if location:
    storage = vyos.commit_archive.URL(location)
    storage.save(remote_file, archive_config_file)
```